src/ripple/rpc/handlers/Tx.cpp

```diff
@@ -97,41 +97,47 @@ Json::Value doTx (RPC::Context& context)
     if (!isHexTxID (txid))
         return rpcError (rpcNOT_IMPL);
 
-    auto txn = context.app.getMasterTransaction ().fetch (
+    auto te = context.app.getMasterTransaction ().fetch (
```

Comment:
One issue we have is updating complete ledgers and updating the database table is not atomic. For example, in online delete, we take the lock for `complete ledgers` update it, release that lock, then take the lock for the database and update the table. We're going to have other issues when we add ledgers to that table as well.

Given that it's not atomic, let me propose two alternatives:

1) what if we saved the complete ledgers both before and after calling `fetch` and used the interesection of those two sets as our returned complete ledgers?

That should make the implementation very easy. Of course, it means a larger interval set may have been searched than what we returned. Given the threading issues with updating the database and updating `complete ledgers` it's going to take a much more extensive fix to actually return the correct ledgers if we use `complete ledgers` for this.

If we do find the transaction, I don't think we should return `complete_ledgers` since we might find a transaction outside of the `complete_ledgers` range.

2) maybe we could use the min and max ledger sequence from the transactions table. Since ledger sequence in indexed, this shouldn't be too expensive.

Reply:
- So per my comments from the cpp-informal thread, can we just acquire the complete ledgers lock first?

- I agree with you about not returning `complete_ledgers` in when the transaction isn't found. Will do.

- I briefly considered using min and max as well, but my concern with that is that (based on the documentation provided for `complete_ledgers`) there may be gaps in the ranges of the completed ledgers stored which wouldn't be captured just by providing a single range based on these values.

Reply:
Went with the first approach.